### ISSUE TITLE
crc setup breaks dns configuration on host - Linux (RHEL9)

### ISSUE THREAD

**Issue by mschindl** [crc-org/crc#3840]
The issue still exists and if it will be installed on a small server then it breaks the dns configuration...

After crc setup my resolv.conf looks like following >>>

$ cat /etc/resolv.conf 
```
# Generated by NetworkManager
search fritz.box local
nameserver 127.0.0.1
options edns0 trust-ad
```

$  cat /etc/os-release
```
NAME="Red Hat Enterprise Linux"
VERSION="9.2 (Plow)"
ID="rhel"
ID_LIKE="fedora"
VERSION_ID="9.2"
PLATFORM_ID="platform:el9"
PRETTY_NAME="Red Hat Enterprise Linux 9.2 (Plow)"
ANSI_COLOR="0;31"
LOGO="fedora-logo-icon"
CPE_NAME="cpe:/o:redhat:enterprise_linux:9::baseos"
HOME_URL="https://www.redhat.com/"
DOCUMENTATION_URL="https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux/9"
BUG_REPORT_URL="https://bugzilla.redhat.com/"

REDHAT_BUGZILLA_PRODUCT="Red Hat Enterprise Linux 9"
REDHAT_BUGZILLA_PRODUCT_VERSION=9.2
REDHAT_SUPPORT_PRODUCT="Red Hat Enterprise Linux"
REDHAT_SUPPORT_PRODUCT_VERSION="9.2
```

Changing dns server for example with "nmtui" doesn't work.

Any ideas on how best to fix this?

_Originally posted by @mschindl in https://github.com/crc-org/crc/issues/1663#issuecomment-1729447033_

**Comment by cfergeau** [comment]
I don't know what does not work/what you are trying to achieve/... so it is hard to answer
resolv.conf points at 127.0.0.1 because NetworkManager is configured to use dnsmasq, and spawned a dnsmasq instance listening on 127.0.0.1. See /etc/NetworkManager if you want to take a closer look at how it's configured.

**Comment by mschindl** [comment]
I noticed this, now I'm trying to reverse it or change it so that DNS queries work again. Until then crc setup doesn't work either, because it can't download packages.

```
INFO Downloading bundle: /home/user/.crc/cache/crc_libvirt_4.13.9_amd64.crcbundle... 
unable to get verified hash for default bundle: Get "https://mirror.openshift.com/pub/openshift-v4/clients/crc/bundles/openshift/4.13.9/sha256sum.txt.sig": dial tcp: lookup mirror.openshift.com: no such host
```

How can I modify "NetworkManager is configured to use dnsmasq, and spawned a dnsmasq instance listening on 127.0.0.1" to use the own dns server as well, like my router?

**Comment by cfergeau** [comment]
> How can I modify "NetworkManager is configured to use dnsmasq, and spawned a dnsmasq instance listening on 127.0.0.1" to use the own dns server as well, like my router?

This should be transparent, is this all configured through dhcp, or are you configuring this manually?

**Comment by mschindl** [comment]
I configured the initial system network configuration manually with "nmtui", no DHCP.
After that I tried to setup crc which breaks the system dns configuration and crc self can't download the required bundle.

**Comment by cfergeau** [comment]
I'd expect nmcli/nmtui to still be usable to configure networking :-/
As an alternative, you can try `crc cleanup && crc config set network-mode user && crc setup`, this should allow crc to work without changing NM configuration.

**Comment by mschindl** [comment]
Great. Looks good so far, but a new error occurs now.

```
INFO Starting CRC VM for openshift 4.13.9...      
failed to expose port :443 -> 192.168.127.2:443: listen tcp :443: bind: address already in use
```
Do you know how to fix that as well?

**Comment by cfergeau** [comment]
`crc config set ingress-https-port 6443` should help. However, you will have to manually change port 443 to 6443 in the openshift URL for the applications you deploy

**Comment by mschindl** [comment]
It did still not work and I received a slightly different error message:
```
INFO Starting CRC VM for openshift 4.13.9...      
failed to expose port :6443 -> 192.168.127.2:443: listen tcp :6443: bind: address already in use
```

**Comment by cfergeau** [comment]
You can take a look with netstat to check what is using this port.

**Comment by mschindl** [comment]
That's why I'm surprised about the issue. 6443 is not in use and should not be.

$ sudo firewall-cmd --list-ports
`631/tcp 5353/udp`

$ sudo lsof -nP -iTCP -sTCP:LISTEN
```
COMMAND     PID USER   FD   TYPE  DEVICE SIZE/OFF NODE NAME
systemd       1 root   67u  IPv4 9071617      0t0  TCP *:111 (LISTEN)
systemd       1 root   69u  IPv6 9071619      0t0  TCP *:111 (LISTEN)
systemd       1 root  159u  IPv6   18794      0t0  TCP *:9090 (LISTEN)
conmon      974 root    5u  IPv4   16159      0t0  TCP *:53 (LISTEN)
conmon      974 root    6u  IPv4   16161      0t0  TCP *:80 (LISTEN)
conmon      974 root    7u  IPv4   16162      0t0  TCP *:443 (LISTEN)
conmon      975 root    5u  IPv4    2693      0t0  TCP *:6789 (LISTEN)
conmon      975 root    6u  IPv4    2694      0t0  TCP *:8080 (LISTEN)
conmon      975 root    7u  IPv4    2695      0t0  TCP *:8081 (LISTEN)
conmon      975 root    8u  IPv4    2696      0t0  TCP *:8443 (LISTEN)
conmon      975 root    9u  IPv4    2697      0t0  TCP *:8843 (LISTEN)
conmon      975 root   10u  IPv4    2698      0t0  TCP *:8880 (LISTEN)
sshd    2072250 root    3u  IPv4 8324512      0t0  TCP *:22 (LISTEN)
sshd    2072250 root    4u  IPv6 8324514      0t0  TCP *:22 (LISTEN)
```

$ netstat -aon | grep 6443
 No output

**Comment by praveenkumar** [comment]
@mschindl Can you try to use a different port because `6443` is used for api server.

```
$ crc config set ingress-https-port 9443
```

**Comment by mschindl** [comment]
I tried it and the following error message appears.

> crc config set ingress-https-port 9443
>> failed to expose port :80 -> 192.168.127.2:80: listen tcp :80: bind: address already in use

<img width="841" alt="image" src="https://github.com/crc-org/crc/assets/26382876/a8f657a4-fff8-4b4f-99f6-e2a1bf604656">

**Comment by cfergeau** [comment]
`80` can be set with `ingress-http-port`, in addition to `ingress-https-port` for port 443

**Comment by mschindl** [comment]
Thank you. Starting works now, but it hangs again:
<img width="905" alt="image" src="https://github.com/crc-org/crc/assets/26382876/d51ca04c-bf0d-40a6-a52c-ca646d506894">

**Comment by praveenkumar** [comment]
@mschindl does it hang here forever or eventually succeed?, if it hangs forever can you provide the logs using `crc start --log-level debug` ?

**Comment by mschindl** [comment]
> @mschindl does it hang here forever or eventually succeed?, if it hangs forever can you provide the logs using `crc start --log-level debug` ?

It hangs here forever and no additional output (I would expect something else) shown. dnsmasq.service failed, but I'm not if it's still needed.

```
INFO 3 operators are progressing: image-registry, kube-storage-version-migrator, service-ca 
DEBU console operator is degraded, Reason: ConsoleCustomRouteSync_FailedDeleteCustomRoutes::DownloadsCustomRouteSync_FailedDeleteCustomRoutes::RouteHealth_FailedRouteGet::SyncLoopRefresh_FailedGet 
DEBU console operator not available, Reason: Deployment_InsufficientReplicas::RouteHealth_RouteNotAdmitted 
DEBU image-registry operator is still progressing, Reason: DeploymentNotCompleted 
DEBU Unexpected operator status for ingress: EvaluationConditionsDetected 
DEBU kube-storage-version-migrator operator is still progressing, Reason: KubeStorageVersionMigrator_Deploying 
DEBU kube-storage-version-migrator operator not available, Reason: KubeStorageVersionMigrator_Deploying 
INFO 2 operators are progressing: image-registry, kube-storage-version-migrator 
DEBU authentication operator not available, Reason: OAuthServerDeployment_NoPod::OAuthServerRouteEndpointAccessibleController_EndpointUnavailable 
DEBU Unexpected operator status for ingress: EvaluationConditionsDetected 
INFO Operator authentication is not yet available 
DEBU authentication operator not available, Reason: OAuthServerDeployment_NoPod 
DEBU Unexpected operator status for ingress: EvaluationConditionsDetected 
INFO Operator authentication is not yet available 
DEBU authentication operator not available, Reason: OAuthServerDeployment_NoPod 
DEBU Unexpected operator status for ingress: EvaluationConditionsDetected 
INFO Operator authentication is not yet available 
DEBU authentication operator not available, Reason: OAuthServerDeployment_NoPod 
DEBU Unexpected operator status for ingress: EvaluationConditionsDetected 
INFO Operator authentication is not yet available 
DEBU authentication operator not available, Reason: OAuthServerDeployment_NoPod 
DEBU Unexpected operator status for ingress: EvaluationConditionsDetected 
INFO Operator authentication is not yet available 
DEBU authentication operator is degraded, Reason: OAuthServerDeployment_UnavailablePod 
DEBU authentication operator not available, Reason: OAuthServerDeployment_NoPod 
DEBU Unexpected operator status for ingress: EvaluationConditionsDetected 
INFO Operator authentication is degraded          
DEBU authentication operator is degraded, Reason: OAuthServerDeployment_UnavailablePod 
DEBU authentication operator not available, Reason: OAuthServerDeployment_NoPod 
DEBU Unexpected operator status for ingress: EvaluationConditionsDetected 
INFO Operator authentication is degraded          
DEBU Unexpected operator status for ingress: EvaluationConditionsDetected 
INFO All operators are available. Ensuring stability... 
DEBU Unexpected operator status for ingress: EvaluationConditionsDetected 
INFO Operators are stable (2/3)...                
DEBU Unexpected operator status for ingress: EvaluationConditionsDetected 
INFO Operators are stable (3/3)...                
DEBU Cluster took 9m0.60545226s to stabilize      
INFO Adding crc-admin and crc-developer contexts to kubeconfig... 


# systemctl status dnsmasq.service
× dnsmasq.service - DNS caching server.
     Loaded: loaded (/usr/lib/systemd/system/dnsmasq.service; disabled; preset: disabled)
     Active: failed (Result: exit-code) since Thu 2023-09-28 10:23:35 CEST; 5h 37min ago
    Process: 3320506 ExecStart=/usr/sbin/dnsmasq (code=exited, status=2)
        CPU: 3ms

Sep 28 10:23:35 server systemd[1]: Starting DNS caching server....
Sep 28 10:23:35 server dnsmasq[3320506]: dnsmasq: failed to create listening socket for 127.0.0.1: Address already in use
Sep 28 10:23:35 server dnsmasq[3320506]: failed to create listening socket for 127.0.0.1: Address already in use
Sep 28 10:23:35 server dnsmasq[3320506]: FAILED to start up
Sep 28 10:23:35 server systemd[1]: dnsmasq.service: Control process exited, code=exited, status=2/INVALIDARGUMENT
Sep 28 10:23:35 server systemd[1]: dnsmasq.service: Failed with result 'exit-code'.
Sep 28 10:23:35 server systemd[1]: Failed to start DNS caching server..
```

**Comment by mschindl** [comment]
An interim status to my issue: CRC is running, but the output from crc start hangs

```
$ crc status
CRC VM:          Running
OpenShift:       Running (v4.13.9)
RAM Usage:       6.252GB of 9.374GB
Disk Usage:      16.14GB of 32.68GB (Inside the CRC VM)
Cache Usage:     20.6GB
Cache Directory: /home/user/.crc/cache

$ oc --kubeconfig=/home/user/.crc/machines/crc/kubeconfig whoami 
system:admin

$ oc --kubeconfig=/home/user/.crc/machines/crc/kubeconfig get nodes
NAME                 STATUS   ROLES                         AGE   VERSION
crc-mmwz7-master-0   Ready    control-plane,master,worker   36d   v1.26.6+6bf3f75
```

**Comment by gbraad** [comment]
Our remote scenario assumes the use of a 'dedicated' server for CRC. It seems in this case port 80, 443, etc are all in use as the machine might be used for several purposes?

**Comment by praveenkumar** [comment]
I had meeting with @mschindl and we were able to move forward around port conflict but at the end when we update the kubeconfig file for dev/admin context, it just hang there. I tried to replicate this on clean RHEL-9 env but not able to produce, suggested to use `strace` to figure out what's blocking this operation. On different terminal we check `crc status` and it was running as expected and with kubeconfig even cluster is accessible.

**Comment by praveenkumar** [comment]
Thanks @mschindl to provide me access to that host and looks like it is hanging during generate token, request is sent to api server and then oauth endpoints which is still using 443 instead what user provide 9443 and this request goes to users pihole setup and hang because there is no response, even `curl 127.0.0.1:443` hangs.

After putting some more info logs this is what happen when it hangs
```
INFO Adding crc-admin and crc-developer contexts to kubeconfig... 
INFO Writing kubeconfig                           
INFO Getting auth ca                              
INFO Getting host info                            
INFO Dialing to api.crc.testing:6443              
INFO Dialing to oauth-openshift.apps-crc.testing:443 
<--- hangs --->
```

**Comment by praveenkumar** [comment]
@mschindl can you again try new version of `crc` on that environment with same setting and let us know if that still hang on adding context part?

**Comment by mschindl** [comment]
Looks good now. Many thanks.

<img width="719" alt="image" src="https://github.com/crc-org/crc/assets/26382876/bef232b2-664f-4814-9d8a-bf65d7fd0e0c">

**Comment by praveenkumar** [comment]
Closing this now

**Comment by mschindl** [comment]
Fixed with version 2.28.0+e1f42d
